tool/planet/agent.go

```diff
@@ -108,151 +104,126 @@ func startLeaderClient(conf *LeaderConfig, errorC chan error) (leaderClient io.C
-		LegacyAPIServerDNSName, "\n")
-	err := ioutil.WriteFile(
-		target,
-		[]byte(contents),
-		SharedFileMask,
-	)
+		LegacyAPIServerDNSName)
+	err := ioutil.WriteFile(target, []byte(contents), SharedFileMask)
 	return trace.Wrap(err)
 }
 
-func updateDNS(conf *LeaderConfig, hostname string, newMasterIP string) error {
-	log.Infof("Setting new leader address to %v in %v", newMasterIP, CoreDNSHosts)
-	err := writeLocalLeader(CoreDNSHosts, newMasterIP)
-	if err != nil {
-		return trace.Wrap(err)
+func updateControlPlaneUnits(localAddr string) handlerFunc {
+	return func(ctx context.Context, _, newLeaderAddr string, logger log.FieldLogger) error {
+		logger.WithField("leader-addr", newLeaderAddr).Info("New leader.")
+		if newLeaderAddr == localAddr {
+			logger.Info("Start control plane units.")
+			updateUnits(ctx, (*unitMonitor).start)
+			return nil
+		}
+		logger.Info("Shut down control plane units.")
+		updateUnits(ctx, (*unitMonitor).stop)
+		return nil
 	}
-	return nil
 }
 
-var electedUnits = []string{
-	"kube-controller-manager.service",
-	"kube-scheduler.service",
-	"kube-apiserver.service",
+func updateElectionParticipation(leaderKey, localAddr string, term time.Duration, client *leader.Client) handlerFunc {
+	return func(ctx context.Context, _, enabledFlag string, logger log.FieldLogger) error {
+		var err error
+		enabled, _ := strconv.ParseBool(enabledFlag)
+		switch enabled {
+		case true:
+			logger.Info("Start election participation.")
+			if err = client.AddVoter(ctx, leaderKey, localAddr, term); err != nil {
+				return trace.Wrap(err)
+			}
+		case false:
+			logger.Info("Stop election participation.")
+		}
+		return nil
+	}
 }
 
-func unitsCommand(command string) error {
-	log.Debugf("executing %v on %v", command, electedUnits)
-	var errors []error
-	for _, unit := range electedUnits {
-		cmd := exec.Command("/bin/systemctl", command, unit)
-		log.Debugf("executing %v", cmd)
-		out, err := cmd.CombinedOutput()
-		if err != nil {
-			errors = append(errors, err)
-			// Instead of failing immediately, complete start of other units
-			log.Warningf("failed to execute %v: %s\n%v", cmd, out, trace.DebugReport(err))
+func updateDNS(_ context.Context, _, newLeaderAddr string, logger log.FieldLogger) error {
+	logger.WithFields(log.Fields{
+		"leader-addr": newLeaderAddr,
+		"file":        CoreDNSHosts,
+	}).Info("Set new leader address.")
+	return writeLocalLeader(CoreDNSHosts, newLeaderAddr)
+}
+
+func updateLoop(ctx context.Context, handler handlerFunc, logger log.FieldLogger) leader.CallbackFn {
+	var localCtx context.Context
+	var localDone <-chan struct{}
+	var cancel context.CancelFunc
+	var wg sync.WaitGroup
+	return func(_, prevValue, newValue string) {
+		select {
+		case <-ctx.Done():
+			return
+		case <-localDone:
+			cancel = nil
+		default:
+			if cancel != nil {
+				localDone = nil
+				cancel()
+				wg.Wait()
+			}
 		}
+		localCtx, cancel = context.WithCancel(ctx)
+		localDone = localCtx.Done()
```

Comment:
Not sure I understand how this works - `localDone` is first defined here but used in the select statement above? Also, I don't see where `localCtx` is used.